File: doqto_backend/app/schemas/group.py
from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, Field

from app.core.constants import GROUP_DESCRIPTION_MAX_LEN, GROUP_NAME_MAX_LEN
from app.core.enums import (
    GroupMemberDmPolicy,
    GroupMemberState,
    GroupPostPolicy,
    GroupRole,
)
from app.schemas.common import ORMModel


class GroupCreateIn(BaseModel):
    name: str = Field(min_length=1, max_length=GROUP_NAME_MAX_LEN)
    description: str | None = Field(default=None, max_length=GROUP_DESCRIPTION_MAX_LEN)
    post_policy: GroupPostPolicy = GroupPostPolicy.ALL_MEMBERS
    member_dm_policy: GroupMemberDmPolicy = GroupMemberDmPolicy.REQUEST
    org_id: uuid.UUID | None = None


class GroupUpdateIn(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=GROUP_NAME_MAX_LEN)
    description: str | None = Field(default=None, max_length=GROUP_DESCRIPTION_MAX_LEN)
    post_policy: GroupPostPolicy | None = None
    member_dm_policy: GroupMemberDmPolicy | None = None
    avatar_url: str | None = Field(default=None, max_length=1024)


class GroupOut(ORMModel):
    id: uuid.UUID
    conversation_id: uuid.UUID
    name: str
    description: str | None
    post_policy: GroupPostPolicy
    member_dm_policy: GroupMemberDmPolicy
    owner_id: uuid.UUID
    org_id: uuid.UUID | None
    avatar_url: str | None
    member_count: int
    created_at: datetime
    updated_at: datetime
    # Populated by the router for the requesting user.
    my_role: GroupRole | None = None
    my_state: GroupMemberState | None = None


class GroupCardOut(BaseModel):
    """A row in "my groups". Groups are invite-only and never discoverable, so
    every card here is one the viewer belongs to or was invited to."""

    id: uuid.UUID
    name: str
    member_count: int
    avatar_url: str | None = None
    description: str | None = None
    my_role: GroupRole | None = None
    my_state: GroupMemberState | None = None


class InviteCreateIn(BaseModel):
    """Either a direct invite (user_id) OR a shareable link (link=true)."""

    user_id: uuid.UUID | None = None
    link: bool = False
    max_uses: int | None = Field(default=None, ge=1)
    expires_at: datetime | None = None


class GroupInviteOut(ORMModel):
    id: uuid.UUID
    group_id: uuid.UUID
    inviter_id: uuid.UUID
    invitee_id: uuid.UUID | None
    token: str | None
    max_uses: int | None
    use_count: int
    expires_at: datetime | None
    state: str
    created_at: datetime


class MemberRoleUpdateIn(BaseModel):
    role: GroupRole


class TransferOwnershipIn(BaseModel):
    user_id: uuid.UUID


class GroupMemberOut(BaseModel):
    user_id: uuid.UUID
    full_name: str
    role: GroupRole
    state: GroupMemberState
    specialty: str | None = None
    avatar_color: str | None = None
    avatar_url: str | None = None


class JoinResultOut(BaseModel):
    """The only self-service way in is an invite link."""

    result: str  # 'joined'
    group_id: uuid.UUID

File: doqto_app/lib/data/models/group.dart
import '../../core/enums/app_enums.dart';
import 'network_profile.dart';

/// Client-only membership-state mirror (backend GroupMemberState). Tolerant:
/// unknown → [active] so a member row never wrongly disappears. NOT parity-gated
/// (backend-only enum — plan §M5).
enum GroupMemberState {
  active,
  banned,
  left,
  removed,
  unknown;

  String get wire => switch (this) {
        GroupMemberState.active => 'active',
        GroupMemberState.banned => 'banned',
        GroupMemberState.left => 'left',
        GroupMemberState.removed => 'removed',
        GroupMemberState.unknown => 'unknown',
      };

  static GroupMemberState fromWire(String? s) => switch (s) {
        'active' => GroupMemberState.active,
        'banned' => GroupMemberState.banned,
        'left' => GroupMemberState.left,
        'removed' => GroupMemberState.removed,
        _ => GroupMemberState.unknown,
      };
}

/// Where this group appeared in "My groups" (client-only tag, set by the state
/// layer from which /me/groups query produced it — the wire has no such field).
enum GroupMembershipTag { member, invited, none }

/// A group. Tolerant hand-written [fromJson] that parses BOTH the full
/// `GroupOut` (detail/create) and the `GroupCardOut` (my groups) — card
/// payloads omit conversation_id/owner_id/post_policy, which default to safe
/// empties. Mirrors backend schemas/group.py.
class Group {
  final String id;
  final String name;
  final String? description;
  final String? memberDmPolicy;
  final String? postPolicy;
  final String? avatarUrl;
  final int memberCount;
  final String conversationId;
  final String ownerId;
  final String? orgId;

  /// The requesting user's role/state, when the payload carries it.
  final GroupRole? myRole;
  final GroupMemberState? myState;
  final DateTime? createdAt;

  /// Client-only tag (see [GroupMembershipTag]); never from the wire.
  final GroupMembershipTag membershipTag;

  const Group({
    required this.id,
    required this.name,
    required this.description,
    required this.memberDmPolicy,
    required this.postPolicy,
    required this.avatarUrl,
    required this.memberCount,
    required this.conversationId,
    required this.ownerId,
    required this.orgId,
    required this.myRole,
    required this.myState,
    required this.createdAt,
    this.membershipTag = GroupMembershipTag.none,
  });

  factory Group.fromJson(Map<String, dynamic> j) => Group(
        id: (j['id'] ?? '') as String,
        name: (j['name'] ?? '') as String,
        description: j['description'] as String?,
        memberDmPolicy: j['member_dm_policy'] as String?,
        postPolicy: j['post_policy'] as String?,
        avatarUrl: j['avatar_url'] as String?,
        memberCount: (j['member_count'] as num?)?.toInt() ?? 0,
        conversationId: (j['conversation_id'] ?? '') as String,
        ownerId: (j['owner_id'] ?? '') as String,
        orgId: j['org_id'] as String?,
        myRole: j['my_role'] != null
            ? GroupRole.fromWire(j['my_role'] as String?)
            : null,
        myState: j['my_state'] != null
            ? GroupMemberState.fromWire(j['my_state'] as String?)
            : null,
        createdAt: j['created_at'] != null
            ? DateTime.tryParse(j['created_at'] as String)
            : null,
      );

  /// True when the requesting user is an active member.
  bool get isMember =>
      myRole != null &&
      (myState == null || myState == GroupMemberState.active);

  /// True when the requesting user can manage the group (owner/admin/mod).
  bool get isAdmin => isMember && (myRole?.isAdminTier ?? false);

  String get initials => initialsOf(name);
  int get avatarIndex => avatarIndexFrom(null, id);

  Group copyWith({
    GroupRole? myRole,
    GroupMemberState? myState,
    int? memberCount,
    GroupMembershipTag? membershipTag,
    bool clearMyRole = false,
  }) =>
      Group(
        id: id,
        name: name,
        description: description,
        memberDmPolicy: memberDmPolicy,
        postPolicy: postPolicy,
        avatarUrl: avatarUrl,
        memberCount: memberCount ?? this.memberCount,
        conversationId: conversationId,
        ownerId: ownerId,
        orgId: orgId,
        myRole: clearMyRole ? null : (myRole ?? this.myRole),
        myState: clearMyRole ? null : (myState ?? this.myState),
        createdAt: createdAt,
        membershipTag: membershipTag ?? this.membershipTag,
      );

  Map<String, dynamic> toCacheJson() => {
        'id': id,
        'name': name,
        'description': description,
        'member_dm_policy': memberDmPolicy,
        'post_policy': postPolicy,
        'avatar_url': avatarUrl,
        'member_count': memberCount,
        'conversation_id': conversationId,
        'owner_id': ownerId,
        'org_id': orgId,
        'my_role': myRole?.wire,
        'my_state': myState?.wire,
        'created_at': createdAt?.toIso8601String(),
      };
}

/// One active member of a group. Mirrors backend GroupMemberOut. Carries NO
/// relationship fields (the members endpoint doesn't return degree/state), so
/// per-row Connect/Message is driven by [relationshipProvider] or a tap that
/// routes to /people/:id.
class GroupMember {
  final String userId;
  final String fullName;
  final String? specialty;
  final String? headline;
  final String? avatarColor;
  final String? avatarUrl;
  final GroupRole role;
  final GroupMemberState state;

  const GroupMember({
    required this.userId,
    required this.fullName,
    required this.specialty,
    required this.headline,
    required this.avatarColor,
    required this.avatarUrl,
    required this.role,
    required this.state,
  });

  factory GroupMember.fromJson(Map<String, dynamic> j) => GroupMember(
        userId: (j['user_id'] ?? '') as String,
        fullName: (j['full_name'] ?? '') as String,
        specialty: j['specialty'] as String?,
        headline: j['headline'] as String?,
        avatarColor: j['avatar_color'] as String?,
        avatarUrl: j['avatar_url'] as String?,
        role: GroupRole.fromWire(j['role'] as String?),
        state: GroupMemberState.fromWire(j['state'] as String?),
      );

  String get initials => initialsOf(fullName);
  int get avatarIndex => avatarIndexFrom(avatarColor, userId);
}

/// Result of accepting an invite link: `{result: 'joined'}`.
class JoinResult {
  final String result;
  final String groupId;
  const JoinResult(this.result, this.groupId);

  bool get joined => result == 'joined';

  factory JoinResult.fromJson(Map<String, dynamic> j) => JoinResult(
        (j['result'] ?? '') as String,
        (j['group_id'] ?? '') as String,
      );
}
